fileManager1API.js

```diff
@@ -0,0 +1,218 @@
+                if (!locationToWindow.has(l)) {
+                    locationToWindow.set(l, new Set());
+                }
+                let window = pathToWindow.get(path);
+                if (window != null) {
+                    locationToWindow.get(l).add(window);
+                }
+            }
+        }
+        this._locationMap = locationToWindow;
+        this.emit('windows-changed');
+    }
+
+    _updateFromXids() {
+        let xidToLocations = this._proxy.XUbuntuOpenLocationsXids;
+        let xidToWindow = getXidToWindow();
+
+        let locationToWindow = new Map();
+        for (let xid in xidToLocations) {
+            let locations = xidToLocations[xid];
+            for (let i = 0; i < locations.length; i++) {
+                let l = locations[i];
+                // Use a set to deduplicate when a window has a
+                // location open in multiple tabs.
+                if (!locationToWindow.has(l)) {
+                    locationToWindow.set(l, new Set());
+                }
+                let window = xidToWindow.get(parseInt(xid));
+                if (window != null) {
+                    locationToWindow.get(l).add(window);
+                }
+            }
+        }
+        this._locationMap = locationToWindow;
+        this.emit('windows-changed');
+    }
+}
+Signals.addSignalMethods(FileManager1Client.prototype);
+
+/**
+ * Construct a map of gtk application window object paths to MetaWindows.
+ */
+function getPathToWindow() {
+    let pathToWindow = new Map();
+
+    for (let i = 0; i < global.workspace_manager.n_workspaces; i++) {
+        let ws = global.workspace_manager.get_workspace_by_index(i);
+        ws.list_windows().map(function(w) {
+            let path = w.get_gtk_window_object_path();
+	    if (path != null) {
+                pathToWindow.set(path, w);
+            }
+        });
+    }
+    return pathToWindow;
+}
+
+/**
+ * Construct a map of XIDs to MetaWindows.
+ *
+ * This is somewhat annoying as you cannot lookup a window by
+ * XID in any way, and must iterate through all of them looking
+ * for a match.
+ */
+function getXidToWindow() {
+    let xidToWindow = new Map();
+
+    for (let i = 0; i < global.workspace_manager.n_workspaces; i++) {
+        let ws = global.workspace_manager.get_workspace_by_index(i);
+        ws.list_windows().map(function(w) {
+            let xid = guessWindowXID(w);
+	    if (xid != null) {
+                xidToWindow.set(parseInt(xid), w);
+            }
+        });
+    }
+    return xidToWindow;
+}
+
+/**
+ * Guesses the X ID of a window.
+ *
+ * This is the basic implementation that is sufficient for Nautilus
+ * windows. The pixel-saver extension has a much more complex
+ * implementation if we ever need it.
+ */
+function guessWindowXID(win) {
+    try {
+        return win.get_description().match(/0x[0-9a-f]+/)[0];
+    } catch (err) {
+        return null;
+    }
+}
+
+var fm1Client = new FileManager1Client();
```

Comment:
Not in love with this, also because it's not destroyed when the extension is disabled, and when not needed.

So in general would be nice to make this part of `DockManager` and then destroy/recreate depending on the locations settings are updated.

I'll prepare something, although a better implementation might be done following what done in https://github.com/micheleg/dash-to-dock/pull/975/commits/4dcaa49ae51f6486b6fad2f9a170f867065d5c1e when that branch will land